### ISSUE TITLE
Uneachable code / Dead code on nested if

### ISSUE THREAD

**Issue by FMorschel** [dart-lang/linter#4150]
**Describe the issue**
There are two lints which could be used for this situation. Most probably `Uneachable code`.

**To Reproduce**
If I have a code similar to this:

```dart
void fn(final bool variable) {
  if (variable == true) {
    print('doing something');
    if (variable == false) {
      print('doing something else');
    }
  }
}
```

**Expected behavior**
I expected the block inside `if (variable == false)` would be warned with `Uneachable code`. I've tried with non-final variables first and I think that could be a case where this is true if I can replace the value of the variable (but shouldn't dart analysis see that that was an `Uneachable code` in that case as well, since I've not updated the value). But with final variables that is not possible at all.

**Comment by lrhn** [comment]
This requires special knowledge about `bool`, to know that there is no value which is equal to both `true` and `false`.

That's not unreasonable, we have and (will) use that information in other places, like switches over `bool` values.
(In that regard, `bool` can be treated like an `enum` with two values with primitive equality, so we know that every `bool` value is equal to precisely one of those two.)

**Comment by srawlins** [comment]
I think this would have been caught (or... there was aspiration that this would have been caught) by the now defunct invariant_booleans linter rule. It became really bogged down with bugs, false positives, false negatives, etc.

I don't see us implementing support for such dead code any time soon.

But I've thought about it: one such neat way to implement it robustly (hopefully) would be if Dart's flow analysis code had hooks and/or an attribute map on which we could hang information. Because implementing checks like this well require flow analysis, but it would be a shame to implement flow analysis n times for n lint rules + language features.
